src/sctools/bam.py

```diff
@@ -307,59 +425,66 @@ def _cleanup(
             f'think about increasing max_mb_per_split.'
         )
 
-    # create all the output files
-    with pysam.AlignmentFile(in_bam, 'rb', check_sq=False) as input_alignments:
+    full_pool = multiprocessing.Pool(num_threads)
 
-        # map files to counts
-        files_to_counts: Dict[pysam.AlignmentFile, int] = {}
-        files_to_names: Dict[pysam.AlignmentFile, str] = {}
-        for i in range(n_subfiles):
-            out_bam_name = out_prefix + '_%d.bam' % i
-            open_bam = pysam.AlignmentFile(
-                out_bam_name, 'wb', template=input_alignments
-            )
-            files_to_counts[open_bam] = 0
-            files_to_names[open_bam] = out_bam_name
+    # Get all the barcodes over all the bams
+    os.write(STDERR, b'Retrieving barcodes from bams\n')
+    result = full_pool.map(
+        partial(get_barcodes_from_bam, tags=tags, raise_missing=raise_missing), in_bams
+    )
 
-        # cycler over files to assign new barcodes to next file
-        file_cycler = cycle(files_to_counts.keys())
+    barcodes_list = list(reduce(lambda set1, set2: set1.union(set2), result))
+    os.write(STDERR, b'Retrieved barcodes from bams\n')
+
+    # Create the barcodes to bin mapping
+    os.write(STDERR, b'Allocating bins\n')
+    barcodes_to_bins_dict = {}
+
+    # barcodes_list will always contain non-None elements from get_barcodes_from_bam
+    if len(barcodes_list) <= n_subfiles:
+        for barcode_index in range(len(barcodes_list)):
+            barcodes_to_bins_dict[barcodes_list[barcode_index]] = barcode_index
+    else:
+        for barcode_index in range(len(barcodes_list)):
+            file_index = barcode_index % n_subfiles
+            barcodes_to_bins_dict[barcodes_list[barcode_index]] = file_index
+
+    # Split the bams by barcode in parallel
+    os.write(STDERR, b'Splitting the bams by barcode\n')
+    # Samtools needs a thread for compression, so we leave half the given threads open.
+    write_pool_threads = math.ceil(num_threads / 2) if num_threads > 2 else 1
+    write_pool = multiprocessing.Pool(write_pool_threads)
+    scattered_split_result = write_pool.map(
+        partial(
+            write_barcodes_to_bins,
+            tags=list(tags),
+            raise_missing=raise_missing,
+            barcodes_to_bins=barcodes_to_bins_dict,
+        ),
+        in_bams,
+    )
 
-        # create an empty map for (tag, barcode) to files
-        tags_to_files = {}
+    bin_indices = list(set(barcodes_to_bins_dict.values()))
+    bins = list(map(lambda index: ["{}_{}".format(out_prefix, index)], bin_indices))
```

Comment:
Can make this a bit shorter:
```suggestion
    bins = list(f"{out_prefix}_{index}" for index in bin_indices)
```